spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -157,8 +159,73 @@ private Schema schemaWithMetadataColumns() {
 
   @Override
   public Scan build() {
-    return new SparkBatchQueryScan(
-        spark, table, readConf, schemaWithMetadataColumns(), filterExpressions);
+    Long snapshotId = readConf.snapshotId();
+    Long asOfTimestamp = readConf.asOfTimestamp();
+
+    Preconditions.checkArgument(snapshotId == null || asOfTimestamp == null,
+        "Cannot set both %s and %s to select which table snapshot to scan",
+        SparkReadOptions.SNAPSHOT_ID, SparkReadOptions.AS_OF_TIMESTAMP);
+
+    Long startSnapshotId = readConf.startSnapshotId();
+    Long endSnapshotId = readConf.endSnapshotId();
+
+    if (snapshotId != null || asOfTimestamp != null) {
+      Preconditions.checkArgument(startSnapshotId == null && endSnapshotId == null,
+          "Cannot set %s and %s for incremental scans when either %s or %s is set",
+          SparkReadOptions.START_SNAPSHOT_ID, SparkReadOptions.END_SNAPSHOT_ID,
+          SparkReadOptions.SNAPSHOT_ID, SparkReadOptions.AS_OF_TIMESTAMP);
+    }
+
+    Preconditions.checkArgument(startSnapshotId != null || endSnapshotId == null,
+        "Cannot set only %s for incremental scans. Please, set %s too.",
+        SparkReadOptions.END_SNAPSHOT_ID, SparkReadOptions.START_SNAPSHOT_ID);
+
+    SparkBatchQueryScan.Context ctx = new SparkBatchQueryScan.Context(
+        snapshotId, asOfTimestamp,
+        startSnapshotId, endSnapshotId);
+
+    return buildBatchQueryScan(ctx);
+  }
+
+  public Scan buildMergeOnReadScan() {
+    Snapshot snapshot = table.currentSnapshot();
```

Comment:
This still needs to validate that the snapshot is not set in SparkReadConf right?

Reply:
Well, I am debating this. If we want to prohibit setting the snapshot id, then we also have to prohibit setting other time-travel properties (3 more). I was not sure about it so I decided to simply ignore them and create a context based on the current snapshot.

Reply:
Added the validation.